### ISSUE TITLE
Proxies to have delegate -> proxies reverse lookup 

### ISSUE THREAD

**Issue by rossbulat** [paritytech/substrate#13596]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Request

We have a bit of a UX nightmare for proxies whereby we cannot determine whether an account is a delegate without also having knowledge of the delegator. This prevents applications to automatically determine a delegate & prevents it from behaving as such.

To solve this we need a reverse lookup from delegate -> proxies to complement proxy -> delegates, so apps can automatically determine what account(s) a delegate is signing transactions for.

We can work around this currently by:
1. Having the delegator address imported in addition to the delegate account (defeats the purpose of having a proxy if you require the delegator to be cold).
2. Allow the user to explicitly submit the address of the delegator so an app can then confirm the state on-chain, & treat the delegate account as a delegate - then utilise local storage to save this state. This approach would break the pattern of importing accounts & things just working.

These options are already being worked on for staking dashboard in order to fulfil our immediate roadmap and move forward on deprecating the controller account for the staking use case specifically.

If this was a trivial feature then I would be comfortable leaving such logic to indexers / a central API to determine delegates. But unstoppable apps should equate to them still being usable even if all central APIs fail. Determining if an account is a delegate is a critical function (we need to know so we can route all calls to `proxy.proxy`, along with the "real" delegator account. This is too critical to rely on centralised APIs IMO.

If proxies are a widespread feature, and we want extensions and the app ecosystem to support them, this reverse lookup solution would be 1 on-chain fix, in comparison to the above solutions whereby every app and extension would implement their own ad-hoc UI / proxy model. Even if we supplied a package to abstract this logic, there is no guarantee every app would use use packages.

For the 20 DOT deposit already committed by the delegator, I feel this added storage is warranted to ease proxy integrations & promote its usage. 

### Are you willing to help with this request?

Yes!

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/proxies-to-have-delegate-proxies-reverse-lookup/2312/1

**Comment by xlc** [comment]
Well, the 20 DOT deposit is calculated by bytes used. You want to insert a new entry and that will increase the deposit further.

Also not sure why this is a critical feature. People will need to know the proxy address upfront anyway. Yes it isn't the best UX but it is manageable. I cannot think any scenario that user want to discover the proxy address.

Take a look another different example: ERC20. You have to import them to wallets or the wallets will use centralized API to discover the tokens. We could put those data onchain but no one have done it for reasons.

**Comment by rossbulat** [comment]
> Well, the 20 DOT deposit is calculated by bytes used. You want to insert a new entry and that will increase the deposit further.

Yes, the 20 DOT is somewhat expensive already so the DOT per bytes could even be decreased.

> Also not sure why this is a critical feature

It is critical if we want widespread support for proxies. People do not need the proxy address at hand if it can be fetched from chain state, which is what we can do if this reverse lookup is introduced.

>  I cannot think any scenario that user want to discover the proxy address.

Staking whereby you import your delegate account via an extension, but not the delegator. We can already perform this discovery process via `staking.bonded` to fetch the controller account, and it is a critical part of auto-discovery within staking dashboard.

Any scenario where a user sets up a proxy for the purpose of not having to import or manage the delegator. Having to copy the public address to tell every single app that an address is a delegate is IMO poor UX & will not stand the test of time as other ecosystems evolve & surpass our UX offerings.

> Take a look another different example: ERC20. You have to import them to wallets or the wallets will use centralized API to discover the tokens. We could put those data onchain but no one have done it for reasons.

But there is no deposit / pay for storage model for this, it is a free for all, just pay a one-time fee to put the contract on-chain. I would not rely on a centralised API for such a critical thing, we do not want to fall into Metamask levels of centralisation.

**Comment by xlc** [comment]
It is not the dApp's job to manage proxy. Ideally, the wallet will have the proxy account and the signer account configuration and expose the proxy account to dApp. On the dApp point of view, it is just a normal account. However, when signing, the wallet will sign `proxy.proxy(tx)` use the signer account.

Otherwise you are asking every single dApp to have logic to support proxy (and multisig), which is a lot of redundant work. You cannot be smart and completely hide it from user. Because a signer account can have multiple proxy so there must be some UI component involved.

**Comment by rossbulat** [comment]
> It is not the dApp's job to manage proxy. 

I disagree here, an app should be able to determine delegates without the user having to jump through hoops to get there. I do not see a use case where users would voluntarily have both the delegate and proxy accounts imported in the same wallet either - I envisage most people will use delegates as a means to not have to use the "real" account, I think this is the main use case of the proxy types. 

Paying 20 DOT and still having to have the proxy account hot will not make sense to many users. The value prop is the fact that the proxy does not need to be in the wallet, & you have a stricter set of calls that the delegate can perform on behalf of that cold account. At least this is how I understand the value proposition of proxies we are discussing here.

> Otherwise you are asking every single dApp to have logic to support proxy (and multisig), which is a lot of redundant work.

This will be the case if we go ahead with this reverse lookup or not - the difference is that the wallet / app can do the work behind the scenes rather than every one implementing their work-arounds of manually asking users to declare the delegate & proxy. The UI components in this case will simply be a list of delegates for each proxy, that can be determined with only 1 query.

**Comment by xlc** [comment]
> apps will most definitely need some level of delegate support.

Yes. And it is infeasible to require every app to build wallet feature to manage accounts. I totally understand your pain when building the staking dashboard. But you need to see that every dApp developer is having the same issue and we need a solution works for everyone.

Also you missing my point that the wallet does not need to require user to import the private key of the proxy account. Just import the pub key is enough.

And this shouldn't be an issue for Ledger. dApp don't talk to Ledger directly, but via a wallet such as Talisman. Then the wallet can translate the tx payload to Ledger. On the Ledger, the user will see this is a `proxy.proxy(tx)` instead of just `tx`, which is expected.

**Comment by rossbulat** [comment]
> But you need to see that every dApp developer is having the same issue and we need a solution works for everyone. Require everyone to build wallet management feature is not the right solution.

Whether we do the reverse lookup or not, this will still be the case. The difference is that only 1 query will be needed to determine the proxy info, versus having to implement various ways to insert public keys and ad-hoc discovery methods that will convolute the ecosystem & ultimately provide a subpar UX compared to automatic discovery - at simply a cost of a few more bytes in storage.

> Also you missing my point that the wallet does not need to require user to import the private key of the proxy account. Just import the pub key is enough.

I totally get it, this is the solution along with looking at existing imported accounts that I am implementing now with my team for the staking dashboard. It is a subpar UX and I can foresee issues not so far down the road when other apps / ecosystem players look to do the same thing.

> And this shouldn't be an issue for Ledger. dApp don't talk to Ledger directly, but via a wallet such as Talisman. Then the wallet can translate the tx payload to Ledger. On the Ledger, the user will see this is a proxy.proxy(tx) instead of just tx, which is expected.

So a large part of my work currently is to move away from web extensions and support native hardware support. This is an optimal experience and dramatically reduces the learning curve of having to import web extensions. Not only this, we are no longer limited to the web browser when web extensions are omitted in favour of hardware wallets.

In response to this I strongly suggest that we do not rely on a future where web extensions or middleware wallets are solely responsible for managing hardware or even facilitating features like proxy. Apps should have this capability to ensure direct communication with hardware wallets is possible.
